stratum/lib/security/credentials_manager.cc

```diff
@@ -9,68 +9,121 @@
     server_credentials_ = ::grpc::InsecureServerCredentials();
   } else {
-    certificate_provider_ = std::make_shared<FileWatcherCertificateProvider>(
-        FLAGS_server_key_file, FLAGS_server_cert_file, FLAGS_ca_cert_file, 1);
-
-    tls_opts_ =
-        std::make_shared<TlsServerCredentialsOptions>(certificate_provider_);
-    tls_opts_->set_cert_request_type(GRPC_SSL_DONT_REQUEST_CLIENT_CERTIFICATE);
-    tls_opts_->watch_root_certs();
-    tls_opts_->watch_identity_key_cert_pairs();
+    auto certificate_provider =
+        std::make_shared<FileWatcherCertificateProvider>(
+            FLAGS_server_key_file, FLAGS_server_cert_file,
+            FLAGS_server_ca_cert_file, kFileRefreshIntervalSeconds);
+    auto tls_opts =
+        std::make_shared<TlsServerCredentialsOptions>(certificate_provider);
+    tls_opts->set_cert_request_type(GRPC_SSL_DONT_REQUEST_CLIENT_CERTIFICATE);
+    tls_opts->watch_root_certs();
+    tls_opts->watch_identity_key_cert_pairs();
+    server_credentials_ = TlsServerCredentials(*tls_opts);
+  }
 
-    server_credentials_ = TlsServerCredentials(*tls_opts_);
+  // Client credentials.
+  if (FLAGS_client_ca_cert_file.empty() && FLAGS_client_key_file.empty() &&
+      FLAGS_client_cert_file.empty()) {
+    client_credentials_ = ::grpc::InsecureChannelCredentials();
+    LOG(WARNING) << "No key files provided, using insecure client credentials!";
+  } else {
+    auto certificate_provider =
+        std::make_shared<FileWatcherCertificateProvider>(
+            FLAGS_client_key_file, FLAGS_client_cert_file,
+            FLAGS_client_ca_cert_file, kFileRefreshIntervalSeconds);
+    auto tls_opts =
+        std::make_shared<TlsChannelCredentialsOptions>(certificate_provider);
+    tls_opts->set_server_verification_option(GRPC_TLS_SERVER_VERIFICATION);
+    tls_opts->watch_root_certs();
+    if (!FLAGS_client_ca_cert_file.empty() && !FLAGS_client_key_file.empty()) {
+      tls_opts->watch_identity_key_cert_pairs();
+    }
+    client_credentials_ = ::grpc::experimental::TlsCredentials(*tls_opts);
   }
+
   return ::util::OkStatus();
 }
 
-::util::Status CredentialsManager::LoadNewCredential(
+::util::Status CredentialsManager::LoadNewServerCredentials(
     const std::string& root_certs, const std::string& cert_chain,
     const std::string& private_key) {
   ::util::Status status;
   // TODO(Kevin): Validate the provided key material if possible
   // TODO(max): According to the API of FileWatcherCertificateProvider, any key
   // and certifcate update must happen atomically. The below code does not
   // guarantee that.
-  status.Update(WriteStringToFile(root_certs, FLAGS_ca_cert_file));
+  status.Update(WriteStringToFile(root_certs, FLAGS_server_ca_cert_file));
   status.Update(WriteStringToFile(cert_chain, FLAGS_server_cert_file));
   status.Update(WriteStringToFile(private_key, FLAGS_server_key_file));
+  absl::SleepFor(absl::Seconds(kFileRefreshIntervalSeconds));
```

Comment:
This is probably fine. However, the convention in gRPC tests seems to be to wait an extra second to avoid a very unlikely race:

https://github.com/grpc/grpc/blob/2d4f3c56001cd1e1f85734b2f7c5ce5f2797c38a/test/core/security/grpc_tls_certificate_provider_test.cc#L316
```suggestion
  absl::SleepFor(absl::Seconds(kFileRefreshIntervalSeconds + 1));
```

Reply:
sounds good, will also update the tests.

lol at the comment above the linked code